### PR TITLE
Add test runner (Doctor) and basic docker ping Healthcheck

### PR DESCRIPTION
**Pull request by fierlion**
### Summary
Add healthcheck runner (Doctor) and Healthcheck; update acs heartbeat handler to trigger a *basic* docker ping healthcheck.

### Implementation details
Please note:
This PR is meant to only add the Doctor.  It has a basic docker ping for demo purposes only.  
As the first commit to the `feature/instance-health` branch, it's meant as a framework to build into.
Next will be a PR to add the true Container Runtime check, including setting check status and timestamps
Following that will be a PR to aggregate the healthcheck statuses into a message to send to TACS.
Then there will be the actual message sending. 

So this is NOT everything.  It's just the Doctor.
 
The Doctor keeps track of an array of Healthchecks (where Healthcheck is an interface).  It exposes the RunTests() method to trigger each of these checks.
The Healthchecks will contain their current and last status as well as timestamps for each; the Doctor will read these to create its ultimate healthcheck message to send on to TACS.

### Testing
built agent and ran to see the healthchecks appear in `/var/log/ecs/ecs-agent.log`:
```
level=info time=2021-05-11T17:26:37Z msg="Received heartbeat from ACS" module=healthcheck.go
level=info time=2021-05-11T17:26:37Z msg="healthcheck response: {0xc0005ec800 <nil>}" module=healthcheck.go
level=info time=2021-05-11T17:26:37Z msg="healthcheck result: true" module=doctor.go
```
also added basic unit tests and ran them:
```
[ec2-user@ip-172-31-32-159 agent]$ go test -count=1 -v -tags unit ./doctor/
=== RUN   TestNewDoctor
--- PASS: TestNewDoctor (0.00s)
=== RUN   TestAddHealthcheck
--- PASS: TestAddHealthcheck (0.00s)
=== RUN   TestOk
--- PASS: TestOk (0.00s)
=== RUN   TestUnmarshalHealthcheckStatus
--- PASS: TestUnmarshalHealthcheckStatus (0.00s)
PASS
```
I'll add more detailed tests once we have more detailed Healthchecks.

New tests cover the changes: yes

### Description for the changelog
Add test runner (Doctor) and basic docker ping Healthcheck.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/dockerclient/dockerapi/docker_client.go [hunk 1]
```diff
@@ -162,6 +162,9 @@ type DockerClient interface {
 	// should be provided for the request.
 	ListContainers(context.Context, bool, time.Duration) ListContainersResponse
 
+	// PingDocker returns the Ping response from Docker's SystemPing API
+	PingDocker(context.Context, time.Duration) PingResponse
```
Comment (angelcar) [inline agent/dockerclient/dockerapi/docker_client.go:166]: nit: this method should probably be just called `Ping` since it's already inside `dockerGoClient`.

File: agent/doctor/doctor.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	return &Doctor{
+		healthchecks: []Healthcheck{},
+	}, nil
+}
+
+func (doc *Doctor) AddHealthcheck(healthcheck Healthcheck) {
+	doc.healthchecks = append(doc.healthchecks, healthcheck)
+}
+
+func (doc *Doctor) RunHealthchecks() bool {
```
Comment (angelcar) [inline agent/doctor/doctor.go:34]: This doesn't seem to be tested

File: agent/doctor/healthcheck.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		client:            client,
+		LatestCheckStatus: HealthcheckStatusInitializing,
+		LatestCheckTime:   time.Now(),
+		LastCheckStatus:   HealthcheckStatusInitializing,
+		LastCheckTime:     time.Now(),
+	}
+}
+
+func (dhc *dockerRuntimeHealthcheck) RunCheck() bool {
+	seelog.Infof("Received heartbeat from ACS")
```
Comment (angelcar) [inline agent/doctor/healthcheck.go:53]: I don't think this log entry belongs here? this health check is/should not aware of ACS, right?

File: agent/doctor/healthcheck.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	// Since is the timestamp when container health status changed
+	LatestCheckTime time.Time `json:"latestStatusTime,omitempty"`
+	// Status is the container health status
+	LastCheckStatus HealthcheckStatus `json:"lastStatus,omitempty"`
+	// Since is the timestamp when container health status changed
+	LastCheckTime time.Time `json:"lastStatusTime,omitempty"`
+	client        dockerapi.DockerClient
+}
+
+func NewDockerRuntimeHealthcheck(client dockerapi.DockerClient) dockerRuntimeHealthcheck {
```
Comment (angelcar) [inline agent/doctor/healthcheck.go:42]: you need to return a pointer here

File: agent/doctor/healthcheck.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/cihub/seelog"
+)
+
+type Healthcheck interface {
+	GetHealthcheckStatus() HealthcheckStatus
+	GetHealthcheckTime() time.Time
+	RunCheck() bool
+}
+
+type dockerRuntimeHealthcheck struct {
```
Comment (angelcar) [inline agent/doctor/healthcheck.go:30]: will this need to be persisted?

File: agent/acs/handler/heartbeat_handler.go [hunk 6]
```diff
@@ -73,7 +79,17 @@ func (heartbeatHandler *heartbeatHandler) handleHeartbeatMessage() {
...
 func (heartbeatHandler *heartbeatHandler) handleSingleHeartbeatMessage(message *ecsacs.HeartbeatMessage) error {
-	// Agent currently has no other action hooked to heartbeat messages, except simple ack
+	// Agent will run healthchecks triggered by ACS heartbeat
+	// healthcheck results will be sent on to TACS
+	dockerHealthcheck := doctor.NewDockerRuntimeHealthcheck(heartbeatHandler.taskEngine.TaskEngineClient())
+	heartbeatHandler.doctor.AddHealthcheck(&dockerHealthcheck)
+
+	go func() {
+		checkResult := heartbeatHandler.doctor.RunHealthchecks()
+		seelog.Infof("handler healthcheck result: %v", checkResult)
```
Comment (shubham2892) [inline agent/acs/handler/heartbeat_handler.go:89]: Do we want to keep this log at info level or debug level as it will be a bit noisy?

File: agent/doctor/doctor.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+	//result will ultimately be an array of something...
+	for _, healthcheck := range doc.healthchecks {
+		res := healthcheck.RunCheck()
+		seelog.Infof("healthcheck result: %v", res)
```
Comment (shubham2892) [inline agent/doctor/doctor.go:38]: Same comment as previous log statement, info or debug level?

File: agent/doctor/doctor.go (hunk 2 above)
Comment (shubham2892) [inline agent/doctor/doctor.go:34]: nit: `RunHealthChecks()` ?

File: agent/doctor/doctor.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	return true
```
Comment (shubham2892) [inline agent/doctor/doctor.go:40]: Right now `RunHealthchecks` always returns `true` irrespective of what we get from individual health checks, do we want to do an `AND` of individual healthcheck and return that?

File: agent/doctor/healthcheck.go (hunk 5 above)
Comment (shubham2892) [reply agent/doctor/healthcheck.go:30]: Across Agent restarts?

File: agent/doctor/healthcheck.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	// Status is the container health status
+	LatestCheckStatus HealthcheckStatus `json:"latestStatus,omitempty"`
+	// Since is the timestamp when container health status changed
+	LatestCheckTime time.Time `json:"latestStatusTime,omitempty"`
+	// Status is the container health status
```
Comment (shubham2892) [inline agent/doctor/healthcheck.go:35]: nit: comments needs to be updated.

File: agent/doctor/healthcheck.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	LastCheckStatus HealthcheckStatus `json:"lastStatus,omitempty"`
+	// Since is the timestamp when container health status changed
```
Comment (shubham2892) [inline agent/doctor/healthcheck.go:37]: nit: comments needs to be updated.

File: agent/doctor/healthcheck.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+	client        dockerapi.DockerClient
+}
+
+func NewDockerRuntimeHealthcheck(client dockerapi.DockerClient) dockerRuntimeHealthcheck {
+	return dockerRuntimeHealthcheck{
+		client:            client,
+		LatestCheckStatus: HealthcheckStatusInitializing,
+		LatestCheckTime:   time.Now(),
+		LastCheckStatus:   HealthcheckStatusInitializing,
+		LastCheckTime:     time.Now(),
```
Comment (shubham2892) [inline agent/doctor/healthcheck.go:48]: can we set this to null instead of initializing it because technically we do should not have these value when initializing?

File: agent/doctor/healthcheck.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	}
+}
+
+func (dhc *dockerRuntimeHealthcheck) RunCheck() bool {
+	seelog.Infof("Received heartbeat from ACS")
+	healthcheckResponse := dhc.client.PingDocker(context.TODO(), time.Second*2)
+	seelog.Infof("healthcheck response: %v", healthcheckResponse)
```
Comment (shubham2892) [inline agent/doctor/healthcheck.go:55]: Should we move this to debug?

File: agent/dockerclient/dockerapi/docker_client.go (hunk 1 above)
Comment (fierlion) [reply agent/dockerclient/dockerapi/docker_client.go:166]: I think I'll call it `SystemPing` --
I wanted to differentiate it from a simple Ping as it is calling a dedicated Docker API.
The fancy name for the API in Docker's docs is "System Ping" https://docs.docker.com/engine/api/v1.41/#operation/SystemPing

File: agent/doctor/doctor.go (hunk 2 above)
Comment (fierlion) [reply agent/doctor/doctor.go:34]: it will be.  rev coming.

File: agent/doctor/doctor.go (hunk 8 above)
Comment (fierlion) [reply agent/doctor/doctor.go:40]: we do want that.  Will make the update.

File: agent/doctor/doctor.go (hunk 2 above)
Comment (fierlion) [reply agent/doctor/doctor.go:34]: I've followed the global convention of calling it a `Healthcheck`.

File: agent/doctor/healthcheck.go (hunk 5 above)
Comment (fierlion) [reply agent/doctor/healthcheck.go:30]: Although I added a LastCheckStatus/Time, we don't currently have a use for it.  And the only reason I can see persistance being necessary is to track our Last state.  

As we've talked about in team discussions -- these stats are really about the very latest stat.
So in case we start up after an agent reset (whether by service restart or instance reboot) and the doctor prematurely sends an aggreagate healthcheck message, the worst case would be that we'd see our defaults in cloudwatch.
```
                LatestCheckStatus: HealthcheckStatusInitializing,
                LatestCheckTime:   time.Now(),
```
Also note that the aggregation and message sending isn't part of the present PR but it should take into account the requirement of always initializing healthchecks where necessary.

So long story short: I don't think persistence is necessary.  Yet.

**Review by shubham2892 — APPROVED** [review]
Was the file `output` intended to be added to the commit? LGTM otherwise.

File: agent/doctor/doctor.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+func (doc *Doctor) AddHealthcheck(healthcheck Healthcheck) {
+	doc.healthchecks = append(doc.healthchecks, healthcheck)
+}
+
+func (doc *Doctor) RunHealthchecks() bool {
+	allChecksResult := []bool{}
+	for _, healthcheck := range doc.healthchecks {
+		res := healthcheck.RunCheck()
+		seelog.Debugf("healthcheck result: %v", res)
```
Comment (sharanyad) [inline agent/doctor/doctor.go:38]: could you add more information to this log statement something like `instance health check result`?

File: agent/doctor/doctor_test.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewDoctor(t *testing.T) {
+	newDoctor, _ := NewDoctor()
+	assert.True(t, len(newDoctor.healthchecks) == 0)
```
Comment (sharanyad) [inline agent/doctor/doctor_test.go:27]: could just use `assert.Len` and below

File: agent/doctor/healthcheck.go [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+		LatestCheckStatus: HealthcheckStatusInitializing,
+		LatestCheckTime:   time.Now(),
+		LastCheckStatus:   HealthcheckStatusInitializing,
+		LastCheckTime:     time.Now(),
+	}
+}
+
+func (dhc *dockerRuntimeHealthcheck) RunCheck() bool {
+	healthcheckResponse := dhc.client.SystemPing(context.TODO(), time.Second*2)
+	seelog.Infof("healthcheck response: %v", healthcheckResponse)
```
Comment (sharanyad) [inline agent/doctor/healthcheck.go:54]: `container runtime health check response`

File: agent/doctor/healthcheck.go [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+	return true
+}
+
+func (dhc *dockerRuntimeHealthcheck) GetHealthcheckStatus() HealthcheckStatus {
+	// need to lock/unlock?
+	return dhc.LatestCheckStatus
+}
+
+func (dhc *dockerRuntimeHealthcheck) GetHealthcheckTime() time.Time {
+	// need to lock/unlock?
```
Comment (sharanyad) [inline agent/doctor/healthcheck.go:64]: always safe to add locks

File: agent/doctor/healthcheck.go [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+		client:            client,
+		LatestCheckStatus: HealthcheckStatusInitializing,
+		LatestCheckTime:   time.Now(),
+		LastCheckStatus:   HealthcheckStatusInitializing,
+		LastCheckTime:     time.Now(),
+	}
+}
+
+func (dhc *dockerRuntimeHealthcheck) RunCheck() bool {
+	healthcheckResponse := dhc.client.SystemPing(context.TODO(), time.Second*2)
```
Comment (sharanyad) [inline agent/doctor/healthcheck.go:53]: could we make the time constant? and pass in `context.Background` if possible?

Comment (fierlion) [reply agent/doctor/healthcheck.go:53]: the systemping here is for demo purposes only.  we have a followup PR to add the true docker runtime check.

File: agent/doctor/doctor.go [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func NewDoctor() (*Doctor, error) {
+	return &Doctor{
+		healthchecks: []Healthcheck{},
+	}, nil
+}
+
+func (doc *Doctor) AddHealthcheck(healthcheck Healthcheck) {
+	doc.healthchecks = append(doc.healthchecks, healthcheck)
```
Comment (sparrc) [inline agent/doctor/doctor.go:31]: this should probably have a write lock

File: agent/doctor/doctor.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (doc *Doctor) RunHealthchecks() bool {
+	allChecksResult := []bool{}
```
Comment (sparrc) [inline agent/doctor/doctor.go:35]: and this function a read lock i think

File: agent/doctor/healthcheck.go [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+	LastCheckTime time.Time `json:"lastStatusTime,omitempty"`
+	client        dockerapi.DockerClient
+	lock          sync.RWMutex
+}
+
+func NewDockerRuntimeHealthcheck(client dockerapi.DockerClient) *dockerRuntimeHealthcheck {
+	return &dockerRuntimeHealthcheck{
+		client:            client,
+		LatestCheckStatus: HealthcheckStatusInitializing,
+		LatestCheckTime:   time.Now(),
```
Comment (sparrc) [inline agent/doctor/healthcheck.go:48]: shouldnt these timestamps be set when the healthchecks are run in RunCheck ?

File: agent/doctor/healthcheck.go [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+	// Since is the timestamp when container health status changed
+	LastCheckTime time.Time `json:"lastStatusTime,omitempty"`
+	client        dockerapi.DockerClient
+	lock          sync.RWMutex
+}
+
+func NewDockerRuntimeHealthcheck(client dockerapi.DockerClient) *dockerRuntimeHealthcheck {
+	return &dockerRuntimeHealthcheck{
+		client:            client,
+		LatestCheckStatus: HealthcheckStatusInitializing,
```
Comment (sparrc) [inline agent/doctor/healthcheck.go:47]: If these are meant to only be read by the public and read-locked functions (GetHealthcheckStatus), then these should not be public.

File: agent/doctor/healthcheck.go [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+func (dhc *dockerRuntimeHealthcheck) RunCheck() bool {
+	healthcheckResponse := dhc.client.SystemPing(context.TODO(), time.Second*2)
+	seelog.Infof("container runtime healthcheck response: %v", healthcheckResponse)
+	return true
+}
+
+func (dhc *dockerRuntimeHealthcheck) GetHealthcheckStatus() HealthcheckStatus {
+	dhc.lock.RLock()
+	defer dhc.lock.RUnlock()
+	return dhc.LatestCheckStatus
```
Comment (sparrc) [inline agent/doctor/healthcheck.go:63]: I think I might be missing something, where are these variables ever set/changed?

File: agent/doctor/healthcheckstatus.go [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+// permissions and limitations under the License.
+
+package doctor
+
+import (
+	"errors"
+	"strings"
+)
+
+const (
```
Comment (sparrc) [inline agent/doctor/healthcheckstatus.go:21]: I think we could just make these statuses string constants instead of using int32 and iota. I think that way we don't need to have all this custom json marshalling either.

File: agent/doctor/doctor.go [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"github.com/cihub/seelog"
+)
+
+type Doctor struct {
+	healthchecks []Healthcheck
+}
+
+func NewDoctor() (*Doctor, error) {
```
Comment (sparrc) [inline agent/doctor/doctor.go:24]: Having a "New" function that returns a Doctor that doesn't do anything doesnt seem like best interface to me. Perhaps the NewDoctor function should automatically call AddHeakthcheck with all the available healthchecks? Personally this seems like a better experience to me than every caller having to manually add each healthcheck even when using the "New" convenience function.

File: agent/acs/handler/heartbeat_handler.go [hunk 25]
```diff
@@ -73,7 +79,19 @@ func (heartbeatHandler *heartbeatHandler) handleHeartbeatMessage() {
 }
 
 func (heartbeatHandler *heartbeatHandler) handleSingleHeartbeatMessage(message *ecsacs.HeartbeatMessage) error {
-	// Agent currently has no other action hooked to heartbeat messages, except simple ack
+	// TestHandlerDoesntLeakGoroutines unit test is failing because of this section
+	// This should all move under a separate healthcheck handler
+
+	// Agent will run healthchecks triggered by ACS heartbeat
+	// healthcheck results will be sent on to TACS
+	//dockerHealthcheck := doctor.NewDockerRuntimeHealthcheck(heartbeatHandler.taskEngine.TaskEngineClient())
```
[the remark this answers is not in the document]
Comment (fierlion) [reply agent/acs/handler/heartbeat_handler.go:87]: this was used in the testing done and will ultimately end up in a separate healthcheck specific handler.  I left it in to show the trigger point used in the description's testing done.

File: agent/doctor/healthcheck.go (hunk 21 above)
Comment (fierlion) [reply agent/doctor/healthcheck.go:47]: these follow the pattern in our api code and allow us to use the json struct field tags.
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/container/container.go#L102-L111

File: agent/doctor/healthcheck.go (hunk 22 above)
Comment (fierlion) [reply agent/doctor/healthcheck.go:63]: this pr is merging into healthcheck feature branch and has a basic "demo" container runtime check.  This PR is specifically meant to add the Doctor.  We have a follow up pr to add the true container runtime check, and to set the statuses and then a third pr to poll these statuses for the message creation and send to TACS.  I'll update the description with this detail.

File: agent/doctor/healthcheck.go (hunk 20 above)
Comment (fierlion) [reply agent/doctor/healthcheck.go:48]: this pr is merging into healthcheck feature branch and has a basic "demo" container runtime check.  This PR is specifically meant to add the Doctor.  We have a follow up pr to add the true container runtime check, and to set the statuses and then a third pr to poll these statuses for the message creation and send to TACS.  I'll update the description with this detail.

File: agent/doctor/healthcheckstatus.go (hunk 23 above)
Comment (fierlion) [reply agent/doctor/healthcheckstatus.go:21]: I know we have opinions on this pattern in agent codebase.  I prefer to follow the existing pattern when sending messages to TACS or FE.
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/container/container.go#L102-L111

Comment (sparrc) [reply agent/doctor/healthcheckstatus.go:21]: but if the existing pattern is not standard or good I'd argue we should differ from it and improve our code

File: agent/doctor/healthcheck.go (hunk 20 above)
Comment (sparrc) [reply agent/doctor/healthcheck.go:48]: this could still be set in RunCheck even if it's demo code so far couldnt it?

File: agent/doctor/doctor.go (hunk 24 above)
Comment (fierlion) [reply agent/doctor/doctor.go:24]: good idea -- I'll update to let it push all the Healthchecks in

File: agent/doctor/healthcheck.go (hunk 21 above)
Comment (sparrc) [reply agent/doctor/healthcheck.go:47]: why would we follow the pattern in our API code? this code is more internal isn't it?

File: agent/dockerclient/dockerapi/docker_client.go [hunk 26]
```diff
@@ -1099,6 +1102,42 @@ func (dg *dockerGoClient) listImages(ctx context.Context) ListImagesResponse {
 	return ListImagesResponse{ImageIDs: imageIDs, RepoTags: imageRepoTags, Error: nil}
 }
 
+func (dg *dockerGoClient) SystemPing(ctx context.Context, timeout time.Duration) PingResponse {
```
Comment (sharanyad) [inline agent/dockerclient/dockerapi/docker_client.go:1105]: needs unit tests in https://github.com/aws/amazon-ecs-agent/blob/master/agent/dockerclient/dockerapi/docker_client_test.go

**Review by sharanyad — APPROVED** [review]
comment on unit testing. could add in a different PR.
